### PR TITLE
Remove stale entries from L0 files when UDT is not persisted

### PR DESCRIPTION
**Pull request by jowlyzhang**
When user-defined timestamps are not persisted, currently we replace the actual timestamp with min timestamp after an entry is output from compaction iterator. Compaction iterator won't be able to help with removing stale entries this way. This PR adds a wrapper iterator `TimestampStrippingIterator` for `MemTableIterator` that does the min timestamp replacement at the memtable iteration step. It is used by flush and can help remove stale entries from landing in L0 files.

Test plan:
Added unit test

### CODE REVIEW

File: db/memtable.cc [hunk 1]
```diff
@@ -613,6 +613,111 @@ InternalIterator* MemTable::NewIterator(
...
+    key_buf_.clear();
+    if (!Valid()) {
+      return;
+    }
+    Slice original_key = iter_->key();
+    ReplaceInternalKeyWithMinTimestamp(&key_buf_, original_key, ts_sz_);
+  }
+
+  size_t ts_sz_;
+  MemTableIterator* iter_;
```
Comment (pdillinger) [inline db/memtable.cc:706]: Can use ScopedArenaPtr 😁

**Review by pdillinger — APPROVED** [review]
I'm slightly concerned about CPU overhead with the intended application, but this is opt-in functionality that can be optimized later based on where the biggest overheads actually are.

Sorry for the slow review!

File: db/column_family_test.cc [hunk 2]
```diff
@@ -3870,6 +3871,48 @@ TEST_F(ManualFlushSkipRetainUDTTest, ManualFlush) {
...
+  for (int version = 0; version < 100; version++) {
+    if (version == 50) {
+      ASSERT_OK(static_cast_with_check<DBImpl>(db_)->TEST_SwitchMemtable(cfd));
+    }
+    ASSERT_OK(
+        Put(0, "foo", EncodeAsUint64(version), "v" + std::to_string(version)));
+  }
+
+  ASSERT_OK(Flush(0));
+  std::vector<LiveFileMetaData> live_files;
```
[the remark this answers is not in the document]
Comment (jowlyzhang) [reply db/column_family_test.cc]: That's a good point, thank you for the suggestion!

**Review by cbi42 — APPROVED** [review]
range deletion part LGTM

**Comment by jowlyzhang** [comment]
@cbi42  Thank you for your help reviewing this PR!

**Pull request closed without merging**